### PR TITLE
Rt 21 fix double payment

### PR DESCRIPTION
**Pull request by agoncharov-reef**

### CODE REVIEW

File: subscriptions/defaults.py [hunk 1]
```diff
@@ -27,3 +27,4 @@
 DEFAULT_SUBSCRIPTIONS_CACHE_NAME = 'subscriptions'
 DEFAULT_SUBSCRIPTIONS_CURRENCY = 'USD'
 DEFAULT_SUBSCRIPTIONS_TRIAL_PERIOD = relativedelta(seconds=0)
+DEFAULT_ABANDON_PENDING_PAYMENTS_AFTER = timedelta(days=3)
```
Comment (kkalinowski-reef) [inline subscriptions/defaults.py]: How about 9 days? Or even better – not abandoning them automatically at all, just listing them periodically e.g. once a day. That would require a human interaction and his "fault" at determining whether a pending payment is to be abandoned.

This is payment. It's on the Paddle side to determine whether it's a success or a failure. And their default subscription refresh points to three retries, three days apart each. And there's no actual explanation on how "pending" works, so maybe leaving it as "pending" is the right thing to do.

File: subscriptions/tasks.py [hunk 2]
```diff
@@ -61,26 +64,30 @@ def _charge_recurring_subscription(
         expiration_date - at,
     )
 
+    # we don't want to try charging if
+    # 1) there is already ANY charge attempt (successful or not) in this charge period
+    # (so if there was ERROR charge in this period, we will try again only in next period)
```
Comment (kkalinowski-reef) [inline subscriptions/tasks.py:69]: Why?

What would be the user flow in this case? Let's say that you try to charge my card and get declined. You won't allow me to recover from this state. Do I need to buy another subscription? Can I? What will happen in the next subscription period?

Also, what with the abandoned subscriptions? If it was abandoned, it's still pending in the Paddle. It still might finish some day, in which case we could do double charging.

File: subscriptions/tasks.py [hunk 3]
```diff
@@ -122,6 +129,19 @@ def _charge_recurring_subscription(
     # to auto-prolong subscription itself
 
 
+def mark_stuck_pending_payments_as_abandoned(older_than: timedelta = DEFAULT_ABANDON_PENDING_PAYMENTS_AFTER):
```
Comment (kkalinowski-reef) [inline subscriptions/tasks.py]: This is not called automatically, right?

Comment (agoncharov-reef) [reply subscriptions/tasks.py]: What do you mean? It's just a task, one could configure celery beat or aws scheduler to run this task

File: subscriptions/tasks.py (hunk 2 above)
Comment (agoncharov-reef) [reply subscriptions/tasks.py:69]: We have a schedule:
```python
DEFAULT_SUBSCRIPTIONS_OFFLINE_CHARGE_ATTEMPTS_SCHEDULE = (
    timedelta(days=-3),
    timedelta(days=-2),
    timedelta(days=-1),
    timedelta(hours=-12),
    timedelta(hours=-3),
    timedelta(hours=-1),
    timedelta(0),
)
```
We try to charge you 3 days ahead, if it fails then we wait until it's 2 days ahead and try again, etc. Can configure to charge try hour if needed.

Well, I thought that if it's abandoned then we consider that Paddle failed and won't complete it ever. But if it does then double charge, yes. I like your proposal.

File: subscriptions/defaults.py (hunk 1 above)
Comment (agoncharov-reef) [reply subscriptions/defaults.py]: No, there is no "default 3 retries" on paddle side, it doesn't apply to one-off charges:
> There is no built-in dunning for the Charges API, so you will need to manage error handling and retries on your own
https://developer.paddle.com/classic/api-reference/23cf86225523f-create-one-off-charge

Anyway I agree to leave it as PENDING.

**Review by kkalinowski-reef — APPROVED** [review]

**Pull request merged**